Merge HealthKit to develop

markyou opened the pull request.

markyou committed:
Improve logging using CocoaLumberjack/Swift instead of naked NSLog
+ Add CocoaLumberjack/Swift pod
+ Convert NSLog to DDlogInfo/Warn/Error/Verbose
+ Control logging globally, currently via DEBUG vs. non-DEBUG, off for non-DEBUG. We may revisit this if we want some logs to be production logs. And/or may consider adding some hidden UI to control logs, similar to what Nutshell does for what backend to connect to.

Other
+ Update RealmSwift and twitter-text pods

markyou committed:
Use Granola serializer as pod

markyou committed:
Rename properties in HealthKitDataSync

Rename some properties in HealthKitDataSync to signify their local db usages. This in prep for adding support for uploading the local db cached HealthKit data to the backend, draining the local db cache upon successful upload of the same.

markyou committed:
More rename refactor - HealthKitDataSync -> HealthKitDataCache

More rename refactor in advance of work to add upload of cached HealthKit data. Essentially change 'sync' to 'cache', and other similar renaming.

markyou committed:
Update HealthKit settings UI per spec
+ Change "Share Health Data" to "Connect to Health"
+ Update HealthKit status display per spec. (Show user full name and date/time of last update. NB: Currently the date/time is the date/time of last observed HealthKit sample written to cache. Eventually this will change, instead, to reflect last date/time of samples uploaded to backend.)

markyou committed:
Fix bug https://trello.com/c/k9s5xryj

Fix bug caused by regression in commit 12857a89f95bcc11dd63cb1823bcd32db8b53615. The default numberOfRows for sections is 0. The numberOfRows for sections for this cell for an account with one group used to be 1, but, the aforementioned commit removed that, except for the HealthKit section, thus resulting in now content showing for other sections, notably Logout and Version.

markyou committed:
Add support to enable/disable logging in production (disabled by default)

Change the "Server" alert to "Settings". Add an action button to the alert to "Disable logging" if logging is enabled, or to "Enable logging until next launch" if logging is disabled. Logging, if enabled, is verbose.
NOTE: Recall that this Debug UI is enabled by tapping four corners on the login view

markyou committed:
Log sample's source and device when logging samples

We now log source and device properties of HKSample when logging samples.

markyou committed:
Add file logger with rolling logs and support for emailing logs

markyou committed:
Persists user's preference for logging

markyou committed:
Flush file logger log before emailing logs. Without this we may get partial logs.

markyou committed:
Fix crash when flushing logs

markyou committed:
Initial draft implementation of CBG data uploader for iOS

markyou committed:
Fix post body for cgb sample(s) upload

markyou committed:
Improve batch upload of CGM data

For now, in Blip Notes, do initial upload (batches of 100) of all cached CGM samples after login, and every two minutes thereafter. Will need to revisit this, and should move more of this functionality into HealthKitUploader to reduce burden on clients.

markyou committed:
Per discussion with jellyfish dev team remove some specific fields for 'upload' and 'cbg' types for iOS uploader

markyou committed:
Make sure we load uploader stats from user defaults when init'ing HealthKitDataUploader

markyou committed:
Bump build number

markyou committed:
Fix issue with initial load of lastUploadCountBloodGlucoseSamples from user defaults as noted in review by gniezen

markyou committed:
Remove Granola. Add metadata to local cache and to sample payload for upload. Update Realm.

markyou committed:
Fix exception serializing metadata JSON object with dates to NSData

markyou committed:
Clear log files on startup

markyou committed:
Only persist new query anchor if there is no error from the anchor query and after result handler successfully returns. Also fix some Swift style issues.

markyou committed:
Make blood glucose HealthKit data cache and upload to service more robust

markyou committed:
Fix issue where we did not transition properly to start observing new HealthKit samples after reading and caching initial historical backlog of samples

markyou committed:
Add migration to clear local db of cached samples, query anchors, and cached/upload stats to avoid having users of Blip Notes (and Nutshell, which also uses this code) to uninstall/reinstall app to ensure a sane state with recent improvements to caching and upload of HealthKit glucose data

markyou committed:
Tweak migration code

lakenyon committed:
Change HK authorization code to support writing to HK as well (needed for Nutshell).

gniezen commented:
@markyou I'm now realizing that it's not really feasible to try and do a code review of the entire PR (with 644 files changed). Maybe I should rather just focus on sections actually creating records and pushing them to the Tidepool API, as it's mostly data integrity we care about in this case. It looks like that would be `HealthKitDataUploader.swift`?

markyou commented:
@gniezen Yes, the huge number of files includes pods that are installed with source checked in, which is a best-practice (unlike npm). Ignore all the pods stuff. Focus on:

HealthKitData.swift
HealthKitDataCache.swift
HealthKitDataManager.swift
HealthKitDataUploader.swift

APIConnect.swift
AppDelegate.swift
LoginViewController.swift

But, yes, primarily just HealthKitDataUploader.swift for review for data integrity concerns.

markyou committed:
Improve uploader
+ Simplify uploader implementation and fix several issues
+ Remove HealthKitData, HealthKitDataCache, Realm
+ Uploader is now a bit more self-contained with simpler interface
+ Fix several issues with batch upload and sample upload JSON, should be closer to spec now

markyou committed:
Improve logging of start/stop reading of samples in uploader

markyou committed:
Don't enable logging in release build unless "LoggingEnabled" user defaults setting is true

markyou committed:
Add system logger (along with Xcode/terminal logger). Don't clear log files on launch so that we can debug/og background delivery/upload of glucose data

markyou committed:
Remove comment

markyou committed:
Add out-of-range annotations

markyou committed:
Minor cleanup

markyou committed:
Fix Zulu time to include milliseconds

gniezen commented on Urchin/HealthKit/HealthKitDataUploader.swift:
```diff
@@ -252,13 +252,40 @@ class HealthKitDataUploader {
+                        [
+                            "annotationCode": annotationCode,
+                            "annotationValue": annotationValue,
+                            "annotationThreshold": annotationThreshold
```
Should be `{"code": annotationCode, "value": annotationValue, "threshold": annotationThreshold}`?

markyou replied:
Yes, it should. Apologies. Was running on fumes after all nighter. Updating now. Apparently jellyfish doesn't validate the annotations schema? Samples with annotations uploaded without issue using my incorrect key names.

markyou committed:
Fix annotations JSON key names per code review

lakenyon committed:
Update to Swift 2.2 syntax.

markyou committed:
Per feedback from review, for uploadId, use the full md5 hash (not first 12 characters) and drop "HealthKit_"

markyou committed:
Last 24 hour upload optimization and improved logging
+ Implement support for uploading last 24 hours of data first, if there has not been a previous successful anchor query read/upload
+ Subclass UIApplication and move CocoaLumberjack init there so that we are guaranteed to get all logs from beginning of app startup

markyou committed:
More work on uploading last 24 hours of CBG data first on initial launch
+ Don't start uploading unless user has previously connected to Health (NB: revisit this when the new slider UI is added)
+ Attempt to fix a few issues with how uploads are started and how we transition from "last 24 hours" to "anchor" and how we handle "read more" case for "anchor"

markyou committed:
Enable logging all the time on launch by default for now. As noted in TODO, probably shouldn't ship to App Store this way.

lakenyon committed:
Quick fix to disable remember-me functionality (does not change to storing token, that is to do still).

markyou committed:
Fix query time range predicate used when applying the "upload last 24 hours first" optimization on initial connection to Health

lakenyon committed:
Provide a reset method for the HealthKit interface to use when switching HealthKit users.

lakenyon committed:
Trello: Only user logged in with DSA accounts should have HealthKit functionality or UI. Integrated this functionality from Nutshell in the form of a new singleton class object, HealthKitConfiguration.

markyou committed:
"Connect to Health" UX improvements, per spec.
+ Only show "Connect to Health" if available for the logged in user (user has own DSA)
+ Use a switch to enable/disable "Connect to Health". NOTE: For existing beta users that have connected to health before, those users will have to still manually turn on the switch to continue uploading new samples. New users won't have this issue.
+ Fix issue with determining whether user has DSA. We were always creating a Patient object for a user even if no "patient" existed in the userDict. Now we don't. All clients already handled nil patient properly. Without this, a user with access to other DSA but not its own would still have the option to "Connect to Health"
+ Only show upload stats if "Connect to Health" is on for the logged in user and some samples have been uploaded
+ When initializing HealthKitUploader state, make sure to reset internal stats if we did reset data for the user due to changing user. (Previously we would retain the previous user's stats.)
+ When initializing HealthKitUploader state due to resetHealthKitUploaderForNewUser make sure to notify clients that stats have changed so the UI can properly reflect the stats for the new user
+ Rename HealthKitConfiguration.showHealthKitUI() to shouldShowHealthKitUI() to reflect that this method doesn't actually show any UI, but, rather, indicates to caller whether it should present any UI
+ Fix crash handling http response codes in refreshToken. (Expected non-empty response document in order to parse code, but, just look at http response status code directly instead now.)
+ Add some additional logging

markyou committed:
Fix a couple issues with HealthKit data authorization
+ Propagate success/error properly
+ Call completion in the the requestAuthorizationToShareTypes closure. (We were prematurely calling completion, thus resulting in too-soon read (or write) of HealthKit data before authorization.)

markyou committed:
+ Simple rename refactor
+ When uploading most recent samples first, query for 10 days, but limit to 288 samples (about 24 hours worth of samples)

markyou committed:
Specify "App Uses Non-Exempt Encryption" as NO in plist

markyou committed:
Add debug menu (four-corner-tap on login screen) to count BG samples (all and Dexcom only) in HealthKit

markyou committed:
Improvements to blood glucose uploader experience
+ Expand the initial upload phase (for first upload for user, or for new uploader version) to two weeks of data
+ The UI now reflects the phase of upload (most recent, or historical) and, for historical phase, indicates the status of samples uploaded through the latest sample date uploaded (i.e. rather than displaying upload time, the most recently uploaded sample time is shown)
+ Subtract duplicate count before adding stats. So, we only report actual new samples uploaded now. (Internally we still keep track of total samples uploaded, including dups, but, that's not exposed to user. It is logged, though.)
+ Fix Trello issue "[blip notes] [iOS] Toggling on "Connect To Health" and authorizing then switching back to Blip Notes, the switch doesn't actually toggle on for a while." ( https://trello.com/c/4EJ4Ql7Q/99-blip-notes-ios-toggling-on-connect-to-health-and-authorizing-then-switching-back-to-blip-notes-the-switch-doesn-t-actually-toggl)
+ Fix Trello issue "[blip notes] [iOS] Toggling off “Connect to Health” then toggling back on doesn’t work (turns back off right away). Toggling one more time then does actually turn on." (https://trello.com/c/FoIGmMET/100-blip-notes-ios-toggling-off-connect-to-health-then-toggling-back-on-doesn-t-work-turns-back-off-right-away-toggling-one-more-tim)
+ Use batch size of 288 (about 1 day of samples) instead of 100. Perf is much better with this

Other
+ Per Apple guidelines, use "OK" rather than "Okay" in button titles

markyou committed:
Remove write permission for blood glucose data

markyou committed:
Attempt to fix Travis CI build. (Requiring Xcode 7.3 now.)

markyou committed:
Implement Trello card "[Urchin] Update messaging for Healthkit sync progress"
+ Lots of changes to implement Trello card "[Urchin] Update messaging for Healthkit sync progress" per https://trello.com/c/YRhgCUvS/101-urchin-update-messaging-for-healthkit-sync-progress

Other related
+ Fix HTTP 400 error issue that could occur when logging out while uploading. We now short circuit upload completion handlers if uploading is stopped while an upload request was outstanding
+ Add "Find date range for blood glucose samples" debug menu item
+ Improve uploader stats
+ Improve logging

markyou committed:
Add support in debug menu for "Email export of HealthKit blood glucose data"
+ Add debug menu "Email export of HealthKit blood glucose data" per Trello card: https://trello.com/c/zr91c8YL/103-urchin-ios-add-a-debug-menu-that-would-dump-healthkit-blood-glucose-data-to-a-csv-file-that-could-be-emailed-like-the-logs
+ Minor refactor of debug menu
+ Minor reorder of method(s) in HealthKitManager
+ Add "Uploaded x of y days" log to uploader stats
NOTE: These debug menu actions that iterate all blood glucose data in HealthKit do queries with no limit. So, for lots of samples (400K+) it can take a minute or two to complete. And, with too many samples, maybe it could run out of memory and crash. (Haven't seen that yet, even with 700K samples. YMMV.)

markyou committed:
Implement Trello card "[Urchin] [iOS] Change copy to "Dexcom data from Health is delayed 3 hours.""
